File: codes/2024.09/4/7.py
from collections import deque

# 二叉树的节点
class TreeNode:
    def __init__(self, val):
        self.val = val
        self.left:TreeNode | None = None
        self.right:TreeNode | None = None

# 多叉树的节点
class Node:
    def __init__(self, val):
        self.val = val
        # 孩子节点
        self.children = []


# 多叉树的前序和后续遍历

# 前序遍历
def preorder(root:Node):
    if root is None:
        return
    print(root.val)
    for child in root.children:
        preorder(child)

# 后序遍历
def postorder(root:Node):
    if root is None:
        return
    ###### 前序位置  #####
    for child in root.children:
        postorder(child)
    ##### 后序位置  #####
    print(root.val)

File: codes/2024/2024.09/1/7.js
var Node = {
  val: 0,
  left: null,
  right: null,
  parent: null,
};

// 二叉树中两个节点的最近公共祖先
var lowestCommonAncestor = function (p, q) {
  // 分别记录 p 和 q 的父节点，用于移动 p 和 q 到根节点
  let p1 = p;
  let p2 = q;

  while (p1 != p2) {
    // 如果 p1 为空，移动到 q 的父节点
    // 如果 p1 不为空，向根节点方向移动，即指针向指向 p1 的父节点
    if (p1 === null) {
      p1 = q;
    } else {
      p1 = p1.parent;
    }

    // 如果 p2 为空，移动到 p 的父节点
    // 如果 p2 不为空，向根节点方向移动，指针向指向 p2 的父节点
    if (p2 === null) {
      p2 = p;
    } else {
      p2 = p2.parent;
    }
  }
};
